cpp/ql/src/experimental/Likely Bugs/OverrunWriteProductFlow.ql

```diff
@@ -78,6 +78,113 @@ predicate isSinkPairImpl(
   )
 }
 
+module ValidState {
+  /**
+   * In the `StringSizeConfig` configuration we use an integer as the flow state for the second
+   * projection of the dataflow graph. The integer represents an offset that is added to the
+   * size of the allocation. For example, given:
+   * ```cpp
+   * char* p = new char[size + 1];
+   * size += 1;
+   * memset(p, 0, size);
+   * ```
+   * the initial flow state is `1`. This represents the fact that `size + 1` is a valid bound
+   * for the size of the allocation pointed to by `p`. After updating the size using `+=`, the
+   * flow state changes to `0`, which represents the fact that `size + 0` is a valid bound for
+   * the allocation.
+   *
+   * So we need to compute a set of valid integers that represent the offset applied to the
+   * size. We do this in two steps:
+   * 1. We first perform the dataflow traversal that the second projection of the product-flow
+   * library will perform, and visit all the places where the size argument is modified.
+   * 2. Once that dataflow traversal is done, we accumulate the offsets added at each places
+   * where the offset is modified (see `validStateImpl`).
+   *
+   * Because we want to guarantee that each place where we modify the offset has a `PathNode`
+   * we "flip" a boolean flow state in each `isAdditionalFlowStep`. This ensures that the node
+   * has a corresponding `PathNode`.
```

Comment:
Why does this help with guaranteeing that there's a path node? Isn't there always a path node?

Reply:
> Isn't there always a path node?

No. In stage 3 of dataflow the paths are "compressed" so that not every `DataFlow::Node` along the path is shown (see [this predicate](https://github.com/github/codeql/blob/main/go/ql/lib/semmle/go/dataflow/internal/DataFlowImpl.qll#LL2143C32-L2143C32) for the details). However, one of the exit conditions for ending a big step is a flow-state change. So by flipping a bit we ensure that there's a flow-state change, and thus the big step relation will be broken up when we reach an `AddInstruction`.

Reply:
Thanks for clearing this up. I should have known, as I was responsible for this behaviour of path nodes being added.